ENH: build fallback lapack_lite with 64-bit integers on 64-bit platforms

pv opened the pull request:
Build the lapack fallback library (used when no LAPACK installed) with 64-bit integer size when building on a 64-bit platform.

Add an indicator flag in `numpy.linalg.lapack_lite` showing whether it was built with 64-bit integers, and use that in the tests (instead of checking whether numpy was linked with external 64-bit BLAS).

Closes: gh-5906

Note that the 64-bit mode is turned on unconditionally, regardless of the `NPY_USE_BLAS_ILP64` setting. I think this is generally OK --- this stuff is fully internal to Numpy, and the change is not visible outside, and the environment flag is anyway only used as a way for the user to control which BLAS library Numpy links against.

NB. the one `s_rnge` is unused function (error reporting for f2c bounds checking, which is turned off here. Just corrected its prototype here to match `f2c.c`.).

pv committed:
DOC: lapack_lite: fix incorrect information in lapack_lite README

pv committed:
ENH: build fallback lapack_lite with 64-bit integers on 64-bit platforms

Build the lapack fallback library (used when no LAPACK installed) with
64-bit integer size when building on a 64-bit platform.

pv committed:
MAINT: better way of indicating whether numpy was built with ILP64 BLAS

charris commented:
Hmm, I originally marked this for backport but changed my mind as it might lead to confusion going forward to change this in 1.18.1. It is probably a bit riskier as well. @pv If you would like to see this in 1.18 let me know. It would be good if the version used was available in  `np.__config__.show()`.

pv committed:
MAINT: indicate in numpy.__config__ when fallback lapack_lite is used

Make numpy.__config__.show() to print information about the fallback
lapack_lite library when it is used.

pv commented:
I agree this doesn't need backport. Added the information to `numpy.__config__`.

eric-wieser commented on numpy/linalg/lapack_lite/f2c.h:
```diff
@@ -8,15 +8,17 @@
 typedef struct { real r, i; } complex;
 typedef struct { doublereal r, i; } doublecomplex;
-typedef int logical;
+typedef CBLAS_INT logical;
```
I'm a little surprised this is the case, but I assume it comes down to the definitions of functions from cblas?

pv replied:
`npy_cblas.h` does not contain functions that have `logical` arguments, and numpy.linalg does not use such functions either. (LAPACK does however contain such functions.)

I think there is no standard for size of `logical` in Fortran. For gfortran, `-fdefault-integer-8` implies also LOGICAL is 8 bytes, and I think that's also true for ifort.

The "canonical" C API LAPACKE defines `lapack_logical` as `lapack_int`, also in its ILP64 configuration. https://github.com/Reference-LAPACK/lapack/blob/f24797e40e44c6d98ea7fa4e36bd1ec6c02c95e5/LAPACKE/include/lapacke_config.h#L46-L56

eric-wieser commented on numpy/linalg/lapack_lite/f2c.h:
```diff
@@ -352,7 +354,7 @@ extern void s_copy(char *, char *, ftnlen, ftnlen);
 extern integer s_rdfe(cilist *);
 extern integer s_rdue(cilist *);
-extern integer s_rnge(char *, integer, char *, integer);
+extern int s_rnge(char *, int, char *, int);
```
Any idea what defines this function? Perhaps the definition should change. (`f2c.c`?)

pv replied:
The definition is like this in `f2c.c`, and the function is unused.

charris commented:
A release note seems appropriate for this.

pv committed:
DOC: add release note for 64-bit lapack_lite

charris commented:
Thanks Pauli.

pv commented:
I realize unfortunately only now the situation is here the same as with openblas64 without symbol suffix, so stuff like `LD_PRELOAD=/usr/lib64/libopenblas.so python3 -mpytest --pyarg numpy` will segfault in this configuration. I'll do follow-up that renames the symbols too...

Numpy could also consider using symbol hiding on Linux (which I think would also fix this), along e.g. the lines Scipy has been using for many years now:
https://github.com/scipy/scipy/blob/47625f4d5c5dccfe1e678509af64aecf56869717/setup.py#L228-L269
